src/role_utils.rs

```diff
@@ -117,6 +119,32 @@ pub struct CommonConfiguration<T> {
     // BTreeMap to keep some order with the cli arguments.
     #[serde(default)]
     pub cli_overrides: BTreeMap<String, String>,
+    #[serde(default)]
+    #[schemars(schema_with = "pod_overrides_schema")]
```

Comment:
I'm still quite new to rust. What does adding this line here cause to happen? It's not quite clear to me how it's making the containers field not mandatory. Does it depend on what the `gen` parameter in `pod_overrides_schema` ends up being?

Reply:
That's bit of magic. https://github.com/Arnavion/k8s-openapi/ takes our custom CRD struct (e.g. TrinoCluster) and automatically generates a CRD schema that can be used for k8s. Which is pretty convenient, as we can model our CRDs in rust code and let k8s-openapi do all the hard work of creating a k8s schema.
To make things even more convenient k8s-openapi offers read-to-use schemas for all the k8s core types (such as PodTemplateSpec).
E.g. the schema for PodSpec is coming from here: https://github.com/Arnavion/k8s-openapi/blob/136f8ae2195f6b72567b3fcce51b5d1572cbfe39/src/v1_27/api/core/v1/pod_spec.rs#L708
The problem is that the schema has containers listed as required here https://github.com/Arnavion/k8s-openapi/blob/136f8ae2195f6b72567b3fcce51b5d1572cbfe39/src/v1_27/api/core/v1/pod_spec.rs#L1208
So we basically take the schema of PodSpec, modify it slightly (remove the required fields), and than use it for our CRD.

Reply:
The `gen`attribute actually doesn't matter too much, I don't even know what purpose it serves ;) We just pass it along to the `PodTemplateSpec::json_schema` function

Reply:
So a few notes here. The CRDs include a [JSON Schema](https://json-schema.org/) representation of the types, which the K8s API server validates the submitted objects against. `kube::CustomResource` uses the [`schemars::JsonSchema`](https://docs.rs/schemars/0.8.12/schemars/trait.JsonSchema.html) trait to generate this portion of the CRD. `#[schemars(schema_with)]` lets us override the schema used for a specific field, in this case to prevent some fields from being mandatory.

Reply:
As for [`gen`](https://docs.rs/schemars/0.8.12/schemars/gen/struct.SchemaGenerator.html), it's machinery used by schemars to carry configuration, and to be able to share certain kinds of global context (like reusing type definitions, though that's not allowed by K8s anyway).